File: src/pentis/defend/langchain_hook.py
"""Pentis Defend middleware for LangChain agents.

LangChain uses class-based AgentMiddleware with wrap_tool_call and
wrap_model_call methods.
"""

from __future__ import annotations

from typing import Any, cast

from pentis.defend.engine import PolicyEngine
from pentis.defend.loader import default_policy
from pentis.defend.models import BLOCKED_MESSAGE, REDACTED_MESSAGE, DefendPolicy


class PentisDefendMiddleware:
    """LangChain AgentMiddleware that enforces Pentis Defend policies.

    Usage::

        from pentis.defend.langchain_hook import PentisDefendMiddleware

        middleware = PentisDefendMiddleware(policy=my_policy)
        agent = create_agent(model=..., tools=..., middleware=[middleware])
    """

    def __init__(self, policy: DefendPolicy | None = None) -> None:
        self.engine = PolicyEngine(policy or default_policy())

    def wrap_tool_call(self, request: Any, handler: Any) -> Any:
        """Intercept tool calls and enforce policy."""
        tool_call: Any = getattr(request, "tool_call", None)
        tool_name: str = ""
        args: dict[str, Any] | None = None
        tool_call_id: str = ""

        if isinstance(tool_call, dict):
            tool_call_dict = cast(dict[str, Any], tool_call)
            tool_name = str(tool_call_dict.get("name", ""))
            raw_args: Any = tool_call_dict.get("args")
            if isinstance(raw_args, dict):
                args = cast(dict[str, Any], raw_args)
            tool_call_id = str(tool_call_dict.get("id", ""))
        elif tool_call is not None:
            tool_name = str(tool_call)

        decision = self.engine.check_tool(tool_name, args)
        if not decision.allowed:
            # Return a ToolMessage indicating the tool was blocked
            try:
                import langchain_core.messages as _lc_msgs  # type: ignore[import-untyped]

                _lc_any: Any = _lc_msgs
                _ToolMessage: type[Any] = _lc_any.ToolMessage
                return _ToolMessage(
                    content=f"{BLOCKED_MESSAGE} {decision.reason}",
                    tool_call_id=tool_call_id,
                )
            except ImportError:
                # Fallback: return a dict if langchain_core is not available
                return {
                    "content": f"{BLOCKED_MESSAGE} {decision.reason}",
                    "tool_call_id": tool_call_id,
                }

        result: Any = handler(request)

        # Check output for sensitive data
        content: str = str(getattr(result, "content", result))
        output_decision = self.engine.check_content(content, is_input=False)
        if not output_decision.allowed:
            if hasattr(result, "content"):
                result.content = REDACTED_MESSAGE

        return result

    def wrap_model_call(self, request: Any, handler: Any) -> Any:
        """Intercept LLM calls and check inputs for injection patterns."""
        messages: list[Any] | None = getattr(request, "messages", None)
        if messages is not None:
            for msg in messages:
                content: str
                if isinstance(msg, dict):
                    msg_dict = cast(dict[str, str], msg)
                    content = msg_dict.get("content", "")
                else:
                    content = str(getattr(msg, "content", msg))
                decision = self.engine.check_content(content, is_input=True)
                if not decision.allowed:
                    # Block the model call by returning a minimal response
                    try:
                        import langchain_core.messages as _lc_msgs  # type: ignore[import-untyped]

                        _lc_any2: Any = _lc_msgs
                        _AIMessage: type[Any] = _lc_any2.AIMessage
                        return _AIMessage(content=f"{BLOCKED_MESSAGE} {decision.reason}")
                    except ImportError:
                        return {"content": f"{BLOCKED_MESSAGE} {decision.reason}"}

        return handler(request)

File: src/defend/langchain-hook.ts
/**
 * Keelson Defend middleware for LangChain agents.
 *
 * LangChain uses class-based middleware with wrapToolCall and wrapModelCall
 * methods. This module provides a middleware class that enforces Keelson
 * Defend policies on tool calls and LLM invocations.
 */

import { PolicyEngine } from './engine.js';
import { defaultPolicy } from './loader.js';
import type { DefendPolicy } from './types.js';
import { BLOCKED_MESSAGE, REDACTED_MESSAGE } from './types.js';

/** Minimal tool call shape from LangChain. */
export interface LangChainToolCall {
  name?: string;
  args?: Record<string, unknown>;
  id?: string;
}

/** Request object for tool call middleware. */
export interface ToolCallRequest {
  toolCall?: LangChainToolCall | Record<string, unknown>;
}

/** Response object with content. */
export interface ContentResponse {
  content: string;
  [key: string]: unknown;
}

/** Request object for model call middleware. */
export interface ModelCallRequest {
  messages?: Array<Record<string, string> | { content?: string }>;
}

/**
 * LangChain middleware that enforces Keelson Defend policies.
 *
 * Usage:
 * ```ts
 * import { KeelsonDefendMiddleware } from 'keelson/defend/langchain-hook.js';
 *
 * const middleware = new KeelsonDefendMiddleware(myPolicy);
 * // Register with LangChain agent middleware pipeline
 * ```
 */
export class KeelsonDefendMiddleware {
  readonly engine: PolicyEngine;

  constructor(policy?: DefendPolicy) {
    this.engine = new PolicyEngine(policy ?? defaultPolicy());
  }

  /** Intercept tool calls and enforce policy. */
  wrapToolCall(
    request: ToolCallRequest,
    handler: (req: ToolCallRequest) => ContentResponse,
  ): ContentResponse | Record<string, unknown> {
    const toolCall = request.toolCall;
    let toolName = '';
    let args: Record<string, unknown> | null = null;
    let toolCallId = '';

    if (toolCall != null && typeof toolCall === 'object') {
      const tc = toolCall as Record<string, unknown>;
      toolName = String(tc.name ?? '');
      if (typeof tc.args === 'object' && tc.args !== null) {
        args = tc.args as Record<string, unknown>;
      }
      toolCallId = String(tc.id ?? '');
    }

    const decision = this.engine.checkTool(toolName, args);
    if (!decision.allowed) {
      return {
        content: `${BLOCKED_MESSAGE} ${decision.reason}`,
        toolCallId,
      };
    }

    const result = handler(request);

    // Check output for sensitive data
    const content = String(result.content ?? result);
    const outputDecision = this.engine.checkContent(content, false);
    if (!outputDecision.allowed && 'content' in result) {
      result.content = REDACTED_MESSAGE;
    }

    return result;
  }

  /** Intercept LLM calls and check inputs for injection patterns. */
  wrapModelCall(
    request: ModelCallRequest,
    handler: (req: ModelCallRequest) => ContentResponse,
  ): ContentResponse | Record<string, unknown> {
    if (request.messages) {
      for (const msg of request.messages) {
        let content: string;
        if (typeof msg === 'object' && msg !== null) {
          content = String((msg as Record<string, unknown>).content ?? '');
        } else {
          content = String(msg);
        }
        const decision = this.engine.checkContent(content, true);
        if (!decision.allowed) {
          return {
            content: `${BLOCKED_MESSAGE} ${decision.reason}`,
          };
        }
      }
    }

    return handler(request);
  }
}
